### ISSUE TITLE
[SR-1667] Erroneous behavior putting a breakpoint on a function that gets optimized out

### ISSUE THREAD

**Issue by NachoSoto** [swiftlang/llvm-project#4401]
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-1667      |
   |Radar             | None         |
   |Original Reporter | @NachoSoto      |
   |Type              | Bug    |

Attachment: [Download](https://user-images.githubusercontent.com/2727770/164961402-cdfb81cf-5f76-44cf-90f4-d2af172b9ca6.gz)

      
<details>
  <summary>Environment</summary>

Xcode 7.3.1 (7D1014)

</details>

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | LLDB for Swift    |
   |Labels            | Bug, DebugInfo        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: e13c861b0a42bc94ed9a24d36d6ad66a

  </details>

**Issue Description:**

Putting a breakpoint in Xcode on a \`private\` function that is never called (which I imagine gets optimized out, even in \`-Onone\`) results in LLDB stopping somewhere else.

This is very confusing and not at all intuitive. I think the correct behavior would be to never stop on that breakpoint at all (since the function is not called).

**Comment by belkadan** [comment]
@adrian-prantl, any insights?

**Comment by adrian-prantl** [comment]
Couple of questions to further triage this:

-   How did you set the breakpoint? (By symbol name, by line+file, ...?)

-   What's the output of "b" in the LLDB console?

-   Do you have any instructions for reproducing the problem? If you could attach a sample project this would be most helpful. If this is not possible, maybe you can attach the .dSYM from the project?

-   Which version of compiler and LLDB are you using?

-   What was the swift compiler invocation for the file containing the function?

**Comment by NachoSoto** [comment]
\> How did you set the breakpoint? (By symbol name, by line+file, ...?)

Clicking on the line number.

\> Do you have any instructions for reproducing the problem? If you could attach a sample project this would be most helpful. If this is not possible, maybe you can attach the .dSYM from the project?

Probably this reproduces the problem:

``` swift
private final class C {
  func foo() {
     print(C.n)
  }

  func bar() {
    print("This method is not used, try to put a breakpoint on this line")
  }

  // and it will probably break here.
  private static let n: Int = 0
}

C().foo()
```

\> Which version of compiler and LLDB are you using?

Whatever's in Xcode 7.3.1 (7D1014) 🙂

\> What was the swift compiler invocation for the file containing the function?

Uhm, whatever Xcode generated. I'm compiling with \`-Onone\`.

**Comment by adrian-prantl** [comment]
Yes, I can confirm that in this case swiftc doesn't generate code for bar and by extension also no debug info for it. This means that the debugger will move the breakpoint forward until the next line that does have code associated with it, which ends up being in a different function in this case.

At -Onone, the function does show up in SIL, but IRGen is dropping it.

**Comment by NachoSoto** [comment]
Cool that explains the bug then 🙂 Moving the breakpoint forward isn't expected behavior though, is it?

**Comment by slavapestov** [comment]
Can we generate a single unreachable instruction somewhere and stick all debug info for deleted functions there?

**Comment by adrian-prantl** [comment]
Short answer: no. At the very least you would need a symbol + at least one (unique) instruction per function for this to work.  
(As a horrible hack it might be possible to fake a gigantic inline tree on that unreachable instruction that contains all dead functions, but that's not the right solution for the problem described in this bug.)
